File: 2610. Convert an Array Into a 2D Array With Conditions.cpp
class Solution {
public:
    vector<vector<int>> findMatrix(vector<int>& nums) {
        unordered_map<int, int> seen;
        vector<vector<int>> ret;

        for (int &num : nums) {
            int row = seen[num];
            seen[num] += 1;

            if (row >= ret.size()) {
                ret.push_back({});
            };

            ret[row].push_back(num);
        };

        return ret;
    }
};

File: 2610. Convert an Array Into a 2D Array With Conditions.py
class Solution:
    # O(n) time and space complexity
    def findMatrix(self, nums: List[int]) -> List[List[int]]:
        seen = {}
        ret = [[]]

        for num in nums:
            row = seen.get(num, 0)
            seen[num] = row + 1

            if row >= len(ret):
                ret.append([num])
            else:
                ret[row].append(num)

        return ret
